### ISSUE TITLE
amigaos4: lettersfall3 distortion

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#70]
Howdy ! :) We again need your help :)

We at least release for the public latest ogles2 and warp3dnova libraries, so ppls can start help us with betatesting, and first issue with which i want to deal with are distrotions in the lettersfall3 game.

There is sources: http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/lettersfall-src_beta2.99a.zip

Issues we have now (i have reports from 5 testers, for all of them it behaves the same, and give the same distortions of textures) , so its not related to my setup only. Also we tested it on different versions of OS components (but always with latest ogles2/warp3dnova), so we at least can rule out other issues, and it probabaly again warp3dnova or ogles2 issue.

There is how it looks like on my setup when i choice "options" or "howtoplay" or "highscores" or start a game (you can see distorted textures of text, they like doubled and shifted randombly):

http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/find_the_bug/lf3_bug1_options.jpg
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/find_the_bug/lf3_bug2_howtoplay.jpg
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/find_the_bug/lf3_bug3_highscores.jpg
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/find_the_bug/lf3_bug4_startgame.jpg

If you can, retest it on your setup again plz. It may looks good when just run it and go one time to options, but then, when you browse them a bit, go to/from few times, then issues appears. 

If you can't reproduce it and does not matter how many times you go to/from all kind of options, then its os4 onl issue, and at moment we stuck as don't know where to start with .. 

Thanks for help !

**Comment by ptitSeb** [comment]
Hey, glad to see some progress here :)

So, I've rechecked letterfall, and yeah, no issues on my side.
Now, I'va made a capture to see how exactly the draws are done.

![image](https://user-images.githubusercontent.com/3250792/45594382-13050b80-b99a-11e8-9e56-10bd7190e695.png)

I see a lot of small drawing. Most drawing are composed of 
```glUniformMatrix4fv(....);
glDrawArrays(GL_TRIANGLE_FAN, 0, 4);```
And some are a bit more complex when there is a texture binding, with a setting each time of GL_LINEAR for GL_MIN_FILTER and GL_MAG_FILTER

But I think the most interesting thing, to dig with OGLES2 or Warp3D drivers dev, is this:
![image](https://user-images.githubusercontent.com/3250792/45594428-cd950e00-b99a-11e8-99ad-5758654458de.png)

There is the creation of texture (probably some SDL_ttf stuff), drawing of the texture, and immediatly after the drawi the deletion of that texture! I guess this is somehow an issue here (testing that theory may be a bit difficult to code, but I can imaging a way to store deletion of texture in a tempory array and delete them only after swap, but that also mean that aglSwapBuffer from gl4es must be used...).

**Comment by kas1e** [comment]
I did search for all "glDel" words in the lettersfall sources, and found them in the 2 files only: interface.cpp and visuals.cpp. There was 2 strings in the interface.cpp and 5 strings in the visuals.cpp. I comment them all out, recompile, and distortions gone, but, with that, i have another issue: freeze when i navigate a bit over the menus.. Strange why so.. mayby it needs to be delete 100% somewhere for game's logic..

**Comment by ptitSeb** [comment]
The freeze is probably because too many texture have be created... Each line of text creat 1 or more textures (depending on the line)...

**Comment by kas1e** [comment]
But should'nt in that case memory be filled up fully before ? I may try to uncomment back one gldeletes after another to see how it behave .. Then once freeze gone and no distortion happens we can try dig in..

**Comment by ptitSeb** [comment]
Not really, this is mainly GPU ressources, not main memory here

**Comment by kas1e** [comment]
Ok, will try it.. But its cleary bug in our warp3d/ogles2, right ? Just want to make bugreport about with some example.. But dunno what problem to make deletion of texture after drawing done.. Too many ?

**Comment by ptitSeb** [comment]
Yes probably on Warp3D side, but maybe on ogles2, not sure. That change will allow you to check if the theory is correct.
Also, with 2 or 3 month between releases, a workaround can be usefull...

**Comment by kas1e** [comment]
With latest commit can't compile gl.c:

src/gl/gl.c: In function 'amiga_post_swap':
src/gl/gl.c:1096: error: 'state' undeclared (first use in this function)
src/gl/gl.c:1096: error: (Each undeclared identifier is reported only once
src/gl/gl.c:1096: error: for each function it appears in.)

And agl.c in this latest form didn't compiles too, it have those errors:

src/agl/agl.c: In function 'aglCreateContext':
src/agl/agl.c:83: error: 'struct OGLES2IFace' has no member named 'aglCreateContext'
src/agl/agl.c: In function 'aglCreateContextTags':
src/agl/agl.c:91: error: 'va_list' undeclared (first use in this function)
src/agl/agl.c:91: error: (Each undeclared identifier is reported only once
src/agl/agl.c:91: error: for each function it appears in.)
src/agl/agl.c:91: error: expected ';' before 'args'
src/agl/agl.c:94: error: variable 'tag' has initializer but incomplete type
src/agl/agl.c:94: warning: implicit declaration of function 'va_arg'
src/agl/agl.c:94: error: 'args' undeclared (first use in this function)
src/agl/agl.c:94: error: expected expression before 'struct'
src/agl/agl.c:94: error: storage size of 'tag' isn't known
src/agl/agl.c:96: error: 'tag' undeclared (first use in this function)
src/agl/agl.c:97: warning: implicit declaration of function 'va_end'
src/agl/agl.c:98: error: 'struct OGLES2IFace' has no member named 'aglCreateContext'

**Comment by kas1e** [comment]
Yeah, they changed a bit indeed (see there "2" added in main functions, that was done because of some bug which was fixed, but to avoid issue with exiting apps he was in needs to make new functinos for future, and keep old ones for old stuff)., there is latest version of includes:

http://kas1e.mikendezign.com/aos4/gl4es/ogles2_1_22_includes.zip

**Comment by ptitSeb** [comment]
Ok thanks.

I tried to fix all error with commit e6942d5d16e7aae336601696a3b6917d850469f0 but I'm not sure the VA_ARG stuff is correct.
I'm haven't seen a difference between `aglCreateContext(...)` and `aglCreateContext2(...)` maybe i missed something.

**Comment by kas1e** [comment]
As far as i know, thee is no differencs betwee old funcitons and "2" ones in terms of api, that all just hidden inside of ogles2.

As for fixes, yeah, gl.c fine, but agl indeed have problems about VA_ARG stuff:

src/agl/agl.c: In function 'aglCreateContextTags':
src/agl/agl.c:94: error: 'VA_LIST' undeclared (first use in this function)
src/agl/agl.c:94: error: (Each undeclared identifier is reported only once
src/agl/agl.c:94: error: for each function it appears in.)
src/agl/agl.c:94: error: expected ';' before 'args'
src/agl/agl.c:95: warning: implicit declaration of function 'VA_START'
src/agl/agl.c:95: error: 'args' undeclared (first use in this function)
src/agl/agl.c:98: error: variable 'tag' has initializer but incomplete type
src/agl/agl.c:98: warning: implicit declaration of function 'VA_ARG'
src/agl/agl.c:98: error: expected expression before 'struct'
src/agl/agl.c:98: error: storage size of 'tag' isn't known
src/agl/agl.c:100: error: 'tag' undeclared (first use in this function)
src/agl/agl.c:101: warning: implicit declaration of function 'VA_END'


I think to avoid problems, we can just remove all together from agl.c/agl.h aglCreateContextTags() function, as i anyway only use aglCreateContext2.. Or i can just ask on amiga forums so someone skilled in those amiga-specific varags functions can give us ready to use code. What you think ?

**Comment by ptitSeb** [comment]
Ok, I'll disable `aglCreateContextTags`, but if you can have the anwser (I may simply missing an include here).

**Comment by kas1e** [comment]
Yeah, workaround works. I still didn't use directly aglswapbuffers from agl.c, but i instead in SDL itself doing that: 

void os4video_GL_SwapBuffers(_THIS)
{
    amiga_pre_swap();
    // Swap the Buffers!
    if(IOGLES2) {
        IOGLES2->aglSwapBuffers();
    }
    amiga_post_swap();  	
}

Which kind of what you mean imho (i.e. necessaity of amiga_pre_swap() / amiga_post_swap() use).

Now it works, didn't freezes, but i think i still can see something.. Probabaly we just a bit side-move from original problem. Now, if i for example go to the game, play a bit, exit, and then press on "new game" again, then, before normal font draws, i can see a fast blink of distored look. I.e. just millisecond, i can see distored version, and then normal one. 

On minigl i didn't have that of course.. Blah.. and how detect that now :(

**Comment by ptitSeb** [comment]
Yeah, calling pre/post swap is good enough.

Strange that there is still some issue. Still, I guess the hypothesis of the texture beeing deleted just after it's draw command is emited not completly spported seems correct.
You should discuss that with Warp3d driver dev (I doubt there is any special traitement on glDeleteTextures in OGLES2).

**Comment by kas1e** [comment]
What is interesting, is that everywhere else (in other games/apps), we also use gldeletetexutes of course, but seems issue only when its so many one aftter another ? Or becaue it simply delete right after drawing, and other apps/games don't do it ?

**Comment by ptitSeb** [comment]
Yeah, this method of creating a temporary texture for every text piece is very uncommon (not very optimized). Most games will either render all letter in an atlas or at least cache the rendered text.

**Comment by kas1e** [comment]
Created bug report for warp3d , and Hans (warp3d author) says that its indeed probably issue with warp3d, as it looks like you can destroy a texture then create a new one that overwrites the old one before the GPU is done with it. 

Our bugreport site migrate, so if you have in bookmarks old one there is new one:
http://www.amiga.org/developer/bugreports/my_view_page.php

Your account of course still valid there, so if you will have anything to say/read about, there is that bugreport: http://www.amiga.org/developer/bugreports/view.php?id=342

**Comment by ptitSeb** [comment]
Ok, I'll keep an eye on the bugtrack.

**Comment by kas1e** [comment]
I recieved fixed version, it works now same like as if i use gl4es workaround ) So, probably workaround can be removed now. Or we can wait when that version of warp3d will become public, and then remove workaround.

**Comment by ptitSeb** [comment]
As you wish. I guess it depends if you want other people to try some gl4es ports or not.
Even if that specific usage (create / delete / create another texture in the same draw command) is rare, it's not unique to LetterFall3, I think I have seen it in other places.

FYI, if you want to disable the workaround temporarily, simple change line 1966 of src/gl/texture.c from `#ifdef AMIGAOS4` to `#if 0 //def AMIGAOS4`

**Comment by ptitSeb** [comment]
So, what is happening in Amiga world?
